google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -75,35 +76,66 @@ Instant instant() {
    * Wrapper around {@code ReadContext} that releases the session to the pool once the call is
    * finished, if it is a single use context.
    */
-  private static class AutoClosingReadContext implements ReadContext {
-    private final ReadContext delegate;
+  private static class AutoClosingReadContext<T extends ReadContext> implements ReadContext {
+    private final Supplier<T> readContextDelegateSupplier;
+    private T readContextDelegate;
     private final PooledSession session;
     private final boolean isSingleUse;
     private boolean closed;
 
     private AutoClosingReadContext(
-        ReadContext delegate, PooledSession session, boolean isSingleUse) {
-      this.delegate = delegate;
+        Supplier<T> delegateSupplier, PooledSession session, boolean isSingleUse) {
+      this.readContextDelegateSupplier = delegateSupplier;
       this.session = session;
       this.isSingleUse = isSingleUse;
+      try {
+        this.readContextDelegate = delegateSupplier.get();
```

Comment:
I would use this.readContextDelegateSupplier

Reply:
I'm not sure I understand what you mean. The variable `this.readContextDelegate` is of type `T extends ReadContext`, so I don't think it should be named `this.readContextDelegateSupplier`. Or do you mean something else?

Reply:
Apologies for not being clear - what I meant was I would have expected this line to be
`this.readContextDelegate = this.readContextDelegateSupplier.get();`
since we just assigned it earlier. Let me know if I'm missing something.